include/xsimd/arch/xsimd_avx512f.hpp

```diff
@@ -740,40 +740,57 @@ namespace xsimd
         }
 
         // gather
-        template <class A, class T,
-                  typename std::enable_if<std::is_same<uint32_t, T>::value || std::is_same<int32_t, T>::value,
-                                          void>::type>
-        inline batch<T, A> gather(T const* src, batch<int32_t, A> const& index,
+        template <class T, class A, class U, detail::enable_sized_integral_t<T, 4> = 0, detail::enable_sized_integral_t<U, 4> = 0>
+        inline batch<T, A> gather(batch<T, A> const&, T const* src, batch<U, A> const& index,
                                   kernel::requires_arch<avx512f>) noexcept
         {
-            return _mm512_i32gather_epi32(index, src, sizeof(T));
+            return _mm512_i32gather_epi32(index, reinterpret_cast<const int32_t*>(src), sizeof(T));
         }
 
-        template <class A, class T,
-                  typename std::enable_if<std::is_same<uint64_t, T>::value || std::is_same<int64_t, T>::value,
-                                          void>::type>
-        inline batch<T, A> gather(T const* src, batch<int64_t, A> const& index,
+        template <class T, class A, class U, detail::enable_sized_integral_t<T, 8> = 0, detail::enable_sized_integral_t<U, 8> = 0>
+        inline batch<T, A> gather(batch<T, A> const&, T const* src, batch<U, A> const& index,
                                   kernel::requires_arch<avx512f>) noexcept
         {
-            return _mm512_i64gather_epi64(index, src, sizeof(T));
+            return _mm512_i64gather_epi64(index, reinterpret_cast<const long long int*>(src), sizeof(T));
```

Comment:
This cast looks suspicious. Looks like you want a cast to `const void *` instead : https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm512_i64gather_epi64&expand=4295&ig_expand=4021

Reply:
@serge-sans-paille yep, GCC and Clang disagree on this. See the current PR checks.

Reply:
Update: it did work :D